source/_posts/2018-01-31-infrastrucutre-as-code-a-beginners-perspective.markdown

```diff
@@ -0,0 +1,122 @@
+
+Oh wait, I forgot, you need two copies (at least) of this infrastructure, one for production and one for staging.
+
+Now, you have two copies of the same infrastructure. You can start deploying your code for that client.
+
+Two months go by and you have another client that needs the same infrastructure, and there you go again, you click around in AWS and get that working.
+
+With IAC you can have a module that does all the heavy lifting for you, you run that code and in a couple of minutes you are good to go. You need to update all the infrastructure of your clients? No problem just update the module, run it and you are good to go!
+
+If you have been doing this by hand in AWS, then you are probably already seeing the benefits of having IAC. But if you don't believe me yet, just read the next case.
+
+### Case 2 - Going crazy with a complex app
+
+You came up with a great idea, a really cool web app, it started off simple, but now you have a lot of clients that rely on that app. In this case one web server and one database are not going to be enough.
+
+You start by adding one load balancer and another web server, you write the process down in a document. Some time goes by and you need to add another server, you follow the process in the document and you are done.
+
+Next month you realize that only one database server it not enough and you add a replica.
+
+Next you have to move some workload out of the main servers, you need a message queue and a server to run workers, you do it by hand, but at this point you forget to update the documents.
+
+Some months go by and you need to add more queues and workers, but since you forgot to add that into a document it is not easy to do it the exact same way, you do it, it works, but now the queues and workers have slightly different configurations, this is called *Configuration drift*.
+
+You decide to add some levels of caching in you app you need Redis/Memcached, so you need to add more servers for those services. You also need to configure S3 to have you static files, a CDN to serve those worldwide, you need Route53 to configure your routes.
+
+By this time the company has grown, you need to create IAM roles and users for the new developers and give them the correct access rights.
+
+You need more reliability, so you need to create more servers in different datacenters, at this point you go nuts and close the company.
+
+The end of this last example was a bit catastrophic, but imagine yourself maintaining that amount of servers by hand, that does not scale up well.
+
+## Should you be doing IAC?
+
+As everything in tech, it depends. I personally can't imagine a project that would not be better by using IAC. So, in my opinion, if you have the chance to use it, give it a try and find out by yourself!
```

Comment:
Maybe in small projects it's too much overhead, this could be a counter example

Reply:
Another example is if you use something like heroku, then this becomes less relevant

Reply:
You can still use it to configure Heroku also hehe -> https://www.terraform.io/docs/providers/heroku/index.html

Reply:
I will mention this in this part